p751toolbox/field_amd64.s

```diff
@@ -1417,11 +1426,367 @@ TEXT ·fp751Mul(SB), $96-24
+	MOVQ	24+M0, DX		\
+	MULXQ	M1+0(SB), T7, T8	\
+	XORQ	AX, AX			\
+	ADCXQ	T4, T7			\
+	ADCXQ	T8, T0			\
+	MULXQ	M1+8(SB), T10, T8	\
+	ADOXQ	T10, T0			\
+	ADCXQ	T8, T1			\
+	MULXQ	M1+16(SB), T8, T4	\
+	ADOXQ	T8, T1			\
+	ADCXQ	T4, T3			\
+	MULXQ	M1+24(SB), T10, T4	\
+	ADCXQ	T4, T6			\
+	MULXQ	M1+32(SB), T8, T4	\
+	ADCXQ	T4, T5			\
+	MULXQ	M1+40(SB), T9, T4	\
+	ADCXQ	T4, T2			\
+	MULXQ	M1+48(SB), DX, T4	\
+	ADCXQ	AX, T4			\
+					\
+	ADOXQ	T10, T3			\
+	ADOXQ	T8, T6			\
+	ADOXQ	T9, T5			\
+	ADOXQ	DX, T2			\
+	ADOXQ	AX, T4
+
+#define mul256x448bmi2(M0, M1, C, T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10) \
+	MOVQ	0+M0, DX		\
+	MULXQ	M1+0(SB), T1, T0	\
+	MULXQ	M1+8(SB), T3, T2	\
+	MOVQ	T1, 0+C			\	// C0_final
+	XORQ	AX, AX			\
+	MULXQ	M1+16(SB), T5, T4	\
+	ADDQ	T3, T0			\
+	ADCQ	T5, T2			\
+	MULXQ	M1+24(SB), T3, T1	\
+	ADCQ	T3, T4			\
+	MULXQ	M1+32(SB), T6, T5	\
+	ADCQ	T6, T1			\
+	MULXQ	M1+40(SB), T7, T3	\
+	ADCQ	T7, T5			\
+	MULXQ	M1+48(SB), T8, T6	\
+	ADCQ	T8, T3			\
+	ADCQ	AX, T6			\
+					\
+	MOVQ	8+M0, DX		\
+	MULXQ	M1+0(SB), T7, T8	\
+	ADDQ	T7, T0			\
+	MOVQ	T0, 8+C			\	// C1_final
+	ADCQ	T8, T2			\
+	MULXQ	M1+8(SB), T8, T7	\
+	MOVQ	T8, 32+C		\
+	ADCQ	T7, T4			\
+	MULXQ	M1+16(SB), T8, T0	\
+	MOVQ	T8, 40+C		\
+	ADCQ	T1, T0			\
+	MULXQ	M1+24(SB), T7, T1	\
+	ADCQ	T5, T1			\
+	MULXQ	M1+32(SB), T8, T5	\
+	ADCQ	T5, T3			\
+	MULXQ	M1+40(SB), T9, T5	\
+	ADCQ	T5, T6			\
+	MULXQ	M1+48(SB), DX, T5	\
+	ADCQ	AX, T5			\
+					\
+	XORQ	AX, AX			\
+	ADDQ	32+C, T2		\
+	ADCQ	40+C, T4		\
+	ADCQ	T7, T0			\
+	ADCQ	T8, T1			\
+	ADCQ	T9, T3			\
+	ADCQ	DX, T6			\
+	ADCQ	AX, T5			\
+					\
+	MOVQ	16+M0, DX		\
+	MULXQ	M1+0(SB), T7, T8	\
+	ADDQ	T7, T2			\
+	MOVQ	T2, 16+C		\	// C2_final
+	ADCQ	T8, T4			\
+	MULXQ	M1+8(SB), T7, T8	\
+	MOVQ	T7, 32+C		\
+	ADCQ	T8, T0			\
+	MULXQ	M1+16(SB), T8, T2	\
+	MOVQ	T8, 40+C		\
+	ADCQ	T2, T1			\
+	MULXQ	M1+24(SB), T7, T2	\
+	ADCQ	T2, T3			\
+	MULXQ	M1+32(SB), T8, T2	\
+	ADCQ	T2, T6			\
+	MULXQ	M1+40(SB), T9, T2	\
+	ADCQ	T2, T5			\
+	MULXQ	M1+48(SB), DX, T2	\
+	ADCQ	AX, T2			\
+					\
+	XORQ	AX, AX			\
+	ADDQ	32+C, T4		\
+	ADCQ	40+C, T0		\
+	ADCQ	T7, T1			\
+	ADCQ	T8, T3			\
+	ADCQ	T9, T6			\
+	ADCQ	DX, T5			\
+	ADCQ	AX, T2			\
+					\
+	MOVQ	24+M0, DX		\
+	MULXQ	M1+0(SB), T7, T8	\
+	ADDQ	T4, T7			\
+	ADCQ	T8, T0			\
+	MULXQ	M1+8(SB), T10, T8	\
+	MOVQ	T10, 32+C		\
+	ADCQ	T8, T1			\
+	MULXQ	M1+16(SB), T8, T4	\
+	MOVQ	T8, 40+C		\
+	ADCQ	T4, T3			\
+	MULXQ	M1+24(SB), T10, T4	\
+	ADCQ	T4, T6			\
+	MULXQ	M1+32(SB), T8, T4	\
+	ADCQ	T4, T5			\
+	MULXQ	M1+40(SB), T9, T4	\
+	ADCQ	T4, T2			\
+	MULXQ	M1+48(SB), DX, T4	\
+	ADCQ	AX, T4			\
+					\
+	XORQ	AX, AX			\
+	ADDQ	32+C, T0		\
+	ADCQ	40+C, T1		\
+	ADCQ	T10, T3			\
+	ADCQ	T8, T6			\
+	ADCQ	T9, T5			\
+	ADCQ	DX, T2			\
+	ADCQ	AX, T4
+
 TEXT ·fp751MontgomeryReduce(SB), $0-16
 
 	MOVQ z+0(FP), REG_P2
 	MOVQ x+8(FP), REG_P1
 
+	// If the MULX instruction is available, a technique from
+	// https://eprint.iacr.org/2017/1015 can be used for a faster Montgomery
+	// reduction.
+	CMPB ·hasBMI2(SB), $0
```

Comment:
Nor do I. I miss `#ifdef` here. I tried to avoid having a function call and to avoid making a compile-time decision (because you lose portable binaries, but maybe that doesn't actually matter). One approach could be to have two `fp751MontgomeryReduce` implementations and set a function pointer once to the correct one.  Maybe with some more macro's to prevent all the duplicate code. This would increase the binary size, but we probably don't care about that.

Reply:
I don't think compile-time decision is right approach. I would rather see it done at runtime, but only once. ``func init()`` - this is called when module is initialized. I would see some function pointer being assigned there.